File: algorithms/data_structures/stack.js
// =====================================================================
// Stack / Стек — LIFO (Last In, First Out)
// ---------------------------------------------------------------------
// Операции:
//   push(x) — O(1)
//   pop()   — O(1)
//   peek()  — O(1)
//   size()  — O(1)
//
// В JS массив уже работает как стек (push/pop). Класс ниже для образования.
// Применение: вызовы функций, отмена операций, обход (DFS),
//             валидация скобок, инфикс→постфикс.
// =====================================================================

class Stack {
    constructor() {
        this.items = [];
    }

    push(item) {
        this.items.push(item);
    }

    pop() {
        if (this.isEmpty()) return undefined;
        return this.items.pop();
    }

    peek() {
        return this.items[this.items.length - 1];
    }

    isEmpty() {
        return this.items.length === 0;
    }

    size() {
        return this.items.length;
    }
}

// --- Пример применения: проверка сбалансированных скобок ---
function isBalanced(s) {
    const pairs = { ')': '(', ']': '[', '}': '{' };
    const stack = new Stack();
    for (const ch of s) {
        if ('([{'.includes(ch)) stack.push(ch);
        else if (')]}'.includes(ch)) {
            if (stack.pop() !== pairs[ch]) return false;
        }
    }
    return stack.isEmpty();
}

const s = new Stack();
s.push(1);
s.push(2);
s.push(3);
console.log(s.pop()); // 3
console.log(s.peek()); // 2

console.log(isBalanced('({[]})')); // true
console.log(isBalanced('({[}])')); // false

module.exports = { Stack, isBalanced };

File: algorithms/data_structures/stack.py
"""
=====================================================================
Stack / Стек — LIFO
---------------------------------------------------------------------
push/pop/peek — O(1)
В Python list работает как стек (append/pop). collections.deque тоже.
=====================================================================
"""


class Stack:
    def __init__(self):
        self._items = []

    def push(self, item):
        self._items.append(item)

    def pop(self):
        if self.is_empty():
            return None
        return self._items.pop()

    def peek(self):
        return self._items[-1] if self._items else None

    def is_empty(self) -> bool:
        return not self._items

    def size(self) -> int:
        return len(self._items)


def is_balanced(s: str) -> bool:
    pairs = {")": "(", "]": "[", "}": "{"}
    stack = Stack()
    for ch in s:
        if ch in "([{":
            stack.push(ch)
        elif ch in ")]}":
            if stack.pop() != pairs[ch]:
                return False
    return stack.is_empty()


if __name__ == "__main__":
    s = Stack()
    s.push(1); s.push(2); s.push(3)
    print(s.pop())   # 3
    print(s.peek())  # 2

    print(is_balanced("({[]})"))  # True
    print(is_balanced("({[}])"))  # False
